File: frontend/src/types/telemetry.ts
/**
 * Mirrors backend/app/models/telemetry.py (M004 PR3, additive): the newest
 * stored telemetry event of one allowlisted type for one fleet asset —
 * today exactly `docker_status` via GET /api/v1/fleet/{id}/docker-status.
 *
 * The payload is schema-flexible on the wire (collectors fail soft), so
 * every payload field below is optional and consumers render "Not reported"
 * when a value is absent — never a fake zero.
 */

/** Per-container facts + stats from the host collector's docker_status payload. */
export interface DockerContainer {
  name?: string;
  image?: string | null;
  status?: string | null;
  exit_code?: number | null;
  restart_count?: number | null;
  started_at?: string | null;
  uptime_seconds?: number | null;
  network_mode?: string | null;
  networks?: string[];
  cpu_percent?: number | null;
  memory_percent?: number | null;
  /** e.g. "132MiB / 3.7GiB" — docker stats' own rendering, shown verbatim. */
  memory_usage?: string | null;
  network_rx_bytes?: number | null;
  network_tx_bytes?: number | null;
}

/** docker_status event payload (collectors/host_pi/docker_stats.py). */
export interface DockerStatusPayload {
  daemon_running?: boolean;
  containers_total?: number;
  containers_running?: number;
  containers_failed?: number;
  restart_count_total?: number;
  containers?: DockerContainer[];
}

/** GET /api/v1/fleet/{id}/docker-status response. */
export interface TelemetrySnapshot<P = DockerStatusPayload> {
  fleet_id: string;
  event_type: string;
  /** Source timestamp (collector clock, UTC). */
  timestamp: string;
  /** Ingestion timestamp stamped by the backend. */
  received_at: string;
  schema_version: number;
  payload: P;
}

File: backend/app/models/telemetry.py
"""Latest-telemetry read models (Mission M004 PR3, additive).

The event stream stores arbitrary typed telemetry per asset; the monitor
consumes it internally, but until M004 no REST route exposed "the newest
event of one type for one asset". :class:`TelemetrySnapshot` is that read
model: the newest stored event of a given type, with its source timestamp and
declared payload schema version, so consumers can reason about freshness and
shape.

Deliberately constrained (supervisor-gated contract): only an explicit
server-side allowlist of event types is ever served — today exactly
``docker_status`` via ``GET /api/v1/fleet/{fleet_id}/docker-status``. The
payload is passed through as stored (rendered inert by consumers, never
interpreted — docs/security.md §9).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.models.event import Event

#: Event type carrying Docker daemon/container telemetry (M003 §10).
DOCKER_STATUS_EVENT_TYPE = "docker_status"

#: Server-side allowlist of telemetry types exposed over REST (M004 PR3
#: supervisor gate): the generic event stream stays internal; only these
#: types get a read route. Extend deliberately, one review at a time.
EXPOSED_TELEMETRY_TYPES: tuple[str, ...] = (DOCKER_STATUS_EVENT_TYPE,)


class TelemetrySnapshot(BaseModel):
    """Newest stored telemetry event of one type for one fleet asset."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    fleet_id: str
    event_type: str
    #: Source timestamp of the event (collector clock, normalized UTC).
    timestamp: datetime
    #: Ingestion timestamp stamped by the backend.
    received_at: datetime
    #: Payload schema version declared by the collector (SD-004).
    schema_version: int
    payload: dict[str, Any]

    @classmethod
    def from_event(cls, fleet_id: str, event: Event) -> TelemetrySnapshot:
        """Project one stored event into the snapshot read model."""
        return cls(
            fleet_id=fleet_id,
            event_type=event.event_type,
            timestamp=event.timestamp,
            received_at=event.received_at,
            schema_version=event.schema_version,
            payload=event.payload,
        )
